### DISCUSSION TITLE
What determines where transient window is shown?

### DISCUSSION THREAD

**Discussion by ville-h** [magit/transient#236] [Support] [answered]
For my setup magit's transient used to show its window at the bottom of the emacs frame. Since update to emacs 28.x magit's transient window appears side-by-side on the right of the original window. I am not sure whether this change comes from upgrade to emacs 28.x, magit or transient update.

Any ideas how to control where the transient window appears?

**Comment by tarsius** [comment]
This can be customized using the option `transient-display-buffer-action`. By default the window is displayed at the bottom of the selected frame.

**Comment by ville-h** [comment]
Alright. Looks like `transient-display-buffer-action` is at its default value according to `describe-variable`:
```
(display-buffer-in-side-window
    (side . bottom)
    (dedicated . t)
    (inhibit-same-window . t)
    (window-parameters
        (no-other-window . t)))
```

I am using an additional package called [shackle](https://depp.brause.cc/shackle/) with a configuration of:
```
    (setq shackle-default-rule '(:same t))
    (setq shackle-rules
        '(
             (" *transient*" :popup t)
        )
    )
```
and the odd behaviour does occur when shackle is enabled. It used to work fine and display transient at the bottom of the window until I upgraded to emacs 28.x. Any ideas why transient in particular is being problematic now?

With shackle disabled transient works fine, but that means every other thing opens windows whichever way based on the alignment of the planets. So that's bit of a no go.

**Reply by tarsius** [reply]
`shackle` overrides how the transient window is displayed. Using `:popup t` does not restore displaying it according to `transient-display-buffer-action`. Instead that causes it to be displayed according to some definition that the author of `shackle` determined to be suitable for many "popups". But it is not suitable for the transient window. Maybe `shackle` can be configured to not change how certain buffers/windows are displayed.

Otherwise `shackle` might fall into the same category as `popwin`, which I have concluded is not compatible with `magit`. I have some vague memory that I have indeed come to that conclusion before, but cannot find anything.

> Any ideas why transient in particular is being problematic now?

Well I don't know what it does, but whether a window is displayed to the right or bottom of the selected window depends on a lot of variables and the size of the selected window and frame. Maybe the value of such a variable changed due to updating Emacs or maybe some calculation changed somewhere.

> [by default] every other thing opens windows whichever way based on the alignment of the planets.

No disagreement there, but IMO simplistic packages such as `popwin` and `shackle` are not the answer either. I am afraid I do not have the time to investigate each such package and try to make them compatible with my packages. Maybe its author can help you.

**Reply by tarsius** [reply]
Or it might be as easy as using this:


>    :align and 'above, 'below, 'left, 'right, or t or a function:
>
> Align a new window at the respective side of the current frame or with the default alignment (customizable with shackle-default-alignment) by splitting the root window. If a function is specified, it is called with zero arguments and must return any of the above alignments.

**Reply by ville-h** [reply]
I don't know how to mark your reply as the answer, because it's one of these "reply" comments. So I'll mark mine below as the answer. Thanks for the help.

**Comment by ville-h** [comment, marked as answer]
Shackle's author pointed the finger elsewhere since shackle's code hadn't changed at the time when I observed the behavioral change. I was just now experimenting and giving shackle's documentation a better read and looks like:
```
    (setq shackle-default-rule '(:same t))
    (setq shackle-rules
        '(
             (" *transient*" :popup t :align 'below)
        )
    )
```
is a working combination.
